### ISSUE TITLE
⚠ emcc fails to compile after including a library (without even using it)

### ISSUE THREAD

**Issue by danbugs** [emscripten-core/emscripten#14691]
Initially, I was building my project with `g++` and it worked fine but, after migrating to `emcc`, it no longer builds. For reference, here's my code:

```
#include <iostream>
#include <emscripten.h>
#include <pqxx/pqxx>

extern "C"
{
    int main()
    {
        std::cout << "Hello, World!" << std::endl;
        return 0;
    }
}
```

Currently, I've narrowed down the issue to the presence of the `#include<pqxx/pqxx>` statement. If I remove it and its' respective include and lib search location flags (i.e., `-I` and `-L`), it compiles.

---

Before migrating to `emcc`, I was building with: 
``` 
g++ -c -Wall -g main.cpp -o bin/main.o -std=c++2a
g++ bin/main.o -lpqxx -lpq -o bin/main -std=c++2a
```
... and it worked.

Now, I'm building with: 
```
emcc -c -Wall -g -I /usr/local/include/ main.cpp -o js/main.o -std=c++2a
emcc js/main.o -L /usr/local/lib/ -L /usr/lib/x86_64-linux-gnu/ -o main.js -std=c++2a
```
...  and, while compiling (i.e., the first emcc command) seems to work w/ just some warnings, linking (i.e., the second `emcc` command) fails.

> Note: I added an appendix below w/ more info on the compilation (i.e., my `Makefile` w/ comments).

Finally, this is the error I get:

```
emcc js/main.o -L /usr/local/lib/ -L /usr/lib/x86_64-linux-gnu/ -o main.js -std=c++2a
wasm-ld: error: unknown file type: dso_handle.o
emcc: error: '/home/danbugs/emsdk/upstream/bin/wasm-ld -o main.wasm js/main.o -L/usr/local/lib/ -L/usr/lib/x86_64-linux-gnu/ -L/home/danbugs/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -ldlmalloc -lc_rt_wasm -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined --strip-debug --export main --export emscripten_stack_get_end --export emscripten_stack_get_free --export emscripten_stack_init --export stackSave --export stackRestore --export stackAlloc --export __wasm_call_ctors --export fflush --export __errno_location --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-table -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024' failed (returned 1)
```

For more info on the error (i.e., the full terminal message post compilation and linking), refer to the appendix. 

All and all, I've got no clue what this error is about and so any help is much appreciated! ;D

### Appendix

1. my `Makefile` w/ comments
```
# run with "make g++"
g++: main.cpp
	g++ -c -Wall -g main.cpp -o bin/main.o -std=c++2a
	g++ bin/main.o -lpqxx -lpq -o bin/main -std=c++2a
# ^^^ this what I ran before starting to migrate to Emscripten

# run with "make emcc"
emcc: main.cpp
	emcc -c -Wall -g -I /usr/local/include/ main.cpp -o js/main.o -std=c++2a
# 					 ^^^ had to add this here because, otherwise, it'd complain it couldn't find "<pqxx/pqxx>"
	emcc js/main.o -L /usr/local/lib/ -L /usr/lib/x86_64-linux-gnu/ -o main.js -std=c++2a

# Legend:
# <pqxx/pqxx> -I /usr/local/include/
# -lpqxx -lpq= -L /usr/local/lib/ -L /usr/lib/x86_64-linux-gnu/
```

2. the full terminal message post compilation and linking
```
danbugs@carol ~/repos/inventar $ make emcc
emcc -c -Wall -g -I /usr/local/include/ main.cpp -o js/main.o -std=c++2a
In file included from main.cpp:4:
In file included from /usr/local/include/pqxx/pqxx:4:
In file included from /usr/local/include/pqxx/blob:4:
In file included from /usr/local/include/pqxx/blob.hxx:33:
In file included from /usr/local/include/pqxx/dbtransaction.hxx:19:
In file included from /usr/local/include/pqxx/transaction_base.hxx:37:
In file included from /usr/local/include/pqxx/row.hxx:20:
/usr/local/include/pqxx/field.hxx:221:9: warning: attribute 'unlikely' has no effect when annotating an 'if constexpr' statement [-Wignored-attributes]
        PQXX_UNLIKELY
        ^~~~~~~~~~~~~
/usr/local/include/pqxx/compiler-public.hxx:118:27: note: expanded from macro 'PQXX_UNLIKELY'
#  define PQXX_UNLIKELY [[unlikely]]
                          ^~~~~~~~
/usr/local/include/pqxx/field.hxx:220:7: note: annotating the 'if constexpr' statement here
      if constexpr (not nullness<T>::has_null)
      ^~~~~~~~~~~~
/usr/local/include/pqxx/field.hxx:497:7: warning: attribute 'unlikely' has no effect when annotating an 'if constexpr' statement [-Wignored-attributes]
      PQXX_UNLIKELY
      ^~~~~~~~~~~~~
/usr/local/include/pqxx/compiler-public.hxx:118:27: note: expanded from macro 'PQXX_UNLIKELY'
#  define PQXX_UNLIKELY [[unlikely]]
                          ^~~~~~~~
/usr/local/include/pqxx/field.hxx:494:5: note: annotating the 'if constexpr' statement here
    if constexpr (nullness<T>::has_null)
    ^~~~~~~~~~~~
In file included from main.cpp:4:
In file included from /usr/local/include/pqxx/pqxx:4:
In file included from /usr/local/include/pqxx/blob:4:
In file included from /usr/local/include/pqxx/blob.hxx:33:
In file included from /usr/local/include/pqxx/dbtransaction.hxx:19:
In file included from /usr/local/include/pqxx/transaction_base.hxx:38:
/usr/local/include/pqxx/stream_from.hxx:341:7: warning: attribute 'unlikely' has no effect when annotating an 'if constexpr' statement [-Wignored-attributes]
      PQXX_UNLIKELY
      ^~~~~~~~~~~~~
/usr/local/include/pqxx/compiler-public.hxx:118:27: note: expanded from macro 'PQXX_UNLIKELY'
#  define PQXX_UNLIKELY [[unlikely]]
                          ^~~~~~~~
/usr/local/include/pqxx/stream_from.hxx:338:5: note: annotating the 'if constexpr' statement here
    if constexpr (nullity::has_null)
    ^~~~~~~~~~~~
3 warnings generated.
emcc js/main.o -L /usr/local/lib/ -L /usr/lib/x86_64-linux-gnu/ -o main.js -std=c++2a
wasm-ld: error: unknown file type: dso_handle.o
emcc: error: '/home/danbugs/emsdk/upstream/bin/wasm-ld -o main.wasm js/main.o -L/usr/local/lib/ -L/usr/lib/x86_64-linux-gnu/ -L/home/danbugs/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -ldlmalloc -lc_rt_wasm -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined --strip-debug --export main --export emscripten_stack_get_end --export emscripten_stack_get_free --export emscripten_stack_init --export stackSave --export stackRestore --export stackAlloc --export __wasm_call_ctors --export fflush --export __errno_location --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-table -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024' failed (returned 1)
make: *** [Makefile:10: emcc] Error 1
````

**Comment by SomeDevHere** [comment]
Emscripten cannot use the system libraries. You need to compile the library in emscripten as well as your code.

**Comment by danbugs** [comment]
@SomeDevHere — First of all, thank you for your response! 

While looking into how to do what you suggested, I found this: https://emscripten.org/docs/compiling/Building-Projects.html#using-libraries and mostly followed it. Considering I already had the libpqxx.a and libpq.a files, I just copied them over to my project and modified my `Makefile` to run the following commands:
```
emcc -c -Wall -g -I /usr/local/include/ main.cpp -o js/main.o -std=c++2a
emcc js/main.o libpq.a libpqxx.a -o js/main.js -std=c++2a
```

The weird `wasm-ld: error: unknown file type: dso_handle.o` error is gone but now I get:
```
emcc js/main.o libpq.a libpqxx.a -o js/main.js -std=c++2a
error: undefined symbol: _ZN4pqxx8internal18demangle_type_nameEPKc (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: __ZN4pqxx8internal18demangle_type_nameEPKc may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/home/danbugs/emsdk/node/14.15.5_64bit/bin/node /home/danbugs/emsdk/upstream/emscripten/src/compiler.js /tmp/tmpd06qxkix.txt' failed (returned 1)
```

Adding the `-s LLD_REPORT_UNDEFINED` flag, I get:
```
emcc js/main.o libpq.a libpqxx.a -o js/main.js -std=c++2a -s LLD_REPORT_UNDEFINED
wasm-ld: error: js/main.o: undefined symbol: pqxx::internal::demangle_type_name(char const*)
wasm-ld: error: js/main.o: undefined symbol: pqxx::internal::demangle_type_name(char const*)
wasm-ld: error: js/main.o: undefined symbol: pqxx::internal::demangle_type_name(char const*)
wasm-ld: error: js/main.o: undefined symbol: pqxx::internal::demangle_type_name(char const*)
emcc: error: '/home/danbugs/emsdk/upstream/bin/wasm-ld -o js/main.wasm js/main.o libpq.a libpqxx.a -L/home/danbugs/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -ldlmalloc -lc_rt_wasm -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined-file=/tmp/tmpe10bew_k.undefined --strip-debug --export main --export emscripten_stack_get_end --export emscripten_stack_get_free --export emscripten_stack_init --export stackSave --export stackRestore --export stackAlloc --export __wasm_call_ctors --export fflush --export __errno_location --export ntohs --export malloc --export htons --export setThrew --export free --export __cxa_is_pointer_type --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-table -z stack-size=5242880 --initial-memory=16777216 --entry=main --max-memory=16777216 --global-base=1024' failed (returned 1)
```` 

Any clue why this happens? 🤔

**Comment by danbugs** [comment]
Alright — I'm guessing it was too naïve of me to assume it'd just work w/ the `.a` files that I already had 😅

The correct way to do this seems to be to go back to the beginning and run:
```
emconfigure ./configure
emmake make
```
... as instructed [here](https://emscripten.org/docs/compiling/Building-Projects.html#using-libraries) in the source code of the libraries I'm using.

Unfortunately, I got an error running `emconfigure ./configure` on `lipqxx` that was related to `libpq`. After that, I went as far as to attempt compiling `libpq` itself w/ Emscripten but that still threw me an error related to the `readline` library sooo... Overall, this feels like a rabbit hole I don't want to get into. 

If you're like me and want to use databases w/ C++ and WASM, the current way to do this seems to use [sql.js](https://github.com/sql-js/sql.js) or files.

I'm closing the issue now because I don't think I'll keep pursuing this.
